src/core/runtime/shard.cc

```diff
@@ -117,21 +117,24 @@ class LegateShardingFunctor : public Legion::ShardingFunctor {
                                 const Domain& launch_space,
                                 const size_t total_shards) override
   {
-    auto lo    = proj_functor_->project_point(launch_space.lo(), launch_space);
-    auto hi    = proj_functor_->project_point(launch_space.hi(), launch_space);
-    auto point = proj_functor_->project_point(p, launch_space);
+    auto lo                 = proj_functor_->project_point(launch_space.lo(), launch_space);
+    auto hi                 = proj_functor_->project_point(launch_space.hi(), launch_space);
+    auto point              = proj_functor_->project_point(p, launch_space);
+    uint32_t task_count     = linearize(lo, hi, hi) + 1;
+    uint32_t proc_count     = end_proc_id_ - start_proc_id_;
+    uint32_t global_proc_id = (linearize(lo, hi, point) * proc_count) / task_count + start_proc_id_;
+    auto shard_id           = global_proc_id / per_node_count_;
 
-    auto shard_id = (linearize(lo, hi, point) + offset_) / per_node_count_ + start_node_id_;
 #ifdef DEBUG_LEGATE
-    assert(start_node_id_ <= shard_id && shard_id < end_node_id_);
+    assert(shard_id < total_shards);
 #endif
     return shard_id;
   }
 
  private:
   LegateProjectionFunctor* proj_functor_;
-  uint32_t start_node_id_;
-  uint32_t end_node_id_;
+  uint32_t start_proc_id_;
+  uint32_t end_proc_id_;
   uint32_t offset_;
```

Comment:
I don't think we need `offset_` anymore.

Reply:
yes, removing it